### PR TITLE
Fixing the definition of relevance to consider context

### PR DESCRIPTION
**Pull request by sunishsheth2009**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/sunishsheth2009/mlflow/pull/10134?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/10134/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 10134
```

</p>
</details>

### What changes are proposed in this pull request?
Fixing the definition of relevance to consider context.
We do include grading_context=['context'] but we never consider context in the prompt.

Using the new prompt, we get better results with human annotations. 
Previous prompt: 15.11627906976744% match
New prompt: 88.37209302325582% match

### How is this PR tested?

- [x] Existing unit/integration tests
- [ ] New unit/integration tests
- [x] Manual tests

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/gateway`: AI Gateway service, Gateway client APIs, third-party Gateway integrations
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/metrics/genai/prompts/v1.py [hunk 1]
```diff
@@ -304,16 +304,23 @@ class AnswerCorrectnessMetric:
...
         "Answer relevance: Please give a score from 1-5 based on the degree of relevance to the "
-        "input, where the lowest and highest scores are defined as follows:"
+        "input based on the provided context, where the lowest and highest scores are "
+        "defined as follows:\n"
         "- Score 1: the output doesn't mention anything about the question or is completely "
         "irrelevant to the input.\n"
+        "- Score 2: the output provides some relevance to the input and outputs one aspect "
+        "of the input correctly using the context.\n"
+        "- Score 3: the output mostly outputs for the input based on the provided context but "
+        "is missing or hallucinating on one critical aspect based on the provided context.\n"
```
Comment (prithvikannan) [inline mlflow/metrics/genai/prompts/v1.py:323]: can we have a score 4 in here?

**Review by prithvikannan — APPROVED** [review]
LGTM with small comment

**Comment by annzhang-db** [comment]
Sorry, I'm confused why we want to do this - we originally introduced relevance (to context) [here](https://github.com/mlflow/mlflow/pull/9724), but intentionally split out metrics into answer relevance (which only considers the input) and faithfulness (which only considers the context) to be more consistent with RAGAS: https://docs.ragas.io/en/latest/concepts/metrics/answer_relevance.html. Were the instructions for human annotators to consider relevance to context? In the case that we are adding this, we are blurring the lines between faithfulness and answer relevance.

**Review by annzhang-db — CHANGES_REQUESTED** [review]
Sorry, put this as a comment earlier - but I'm confused why we want to do this. We originally introduced relevance (to context) [here](https://github.com/mlflow/mlflow/pull/9724), but intentionally split out metrics into answer relevance (which only considers the input) and faithfulness (which only considers the context) to be more consistent with RAGAS: https://docs.ragas.io/en/latest/concepts/metrics/answer_relevance.html. Were the instructions for human annotators to consider relevance to context (the definition may be different than what we are looking for with this metric)? In the case that we are adding this, we are blurring the lines between faithfulness and answer relevance and it gets more ambiguous.

**Comment by sunishsheth2009** [comment]
> Sorry, put this as a comment earlier - but I'm confused why we want to do this. We originally introduced relevance (to context) [here](https://github.com/mlflow/mlflow/pull/9724), but intentionally split out metrics into answer relevance (which only considers the input) and faithfulness (which only considers the context) to be more consistent with RAGAS: https://docs.ragas.io/en/latest/concepts/metrics/answer_relevance.html. Were the instructions for human annotators to consider relevance to context (the definition may be different than what we are looking for with this metric)? In the case that we are adding this, we are blurring the lines between faithfulness and answer relevance and it gets more ambiguous.

Yup I thought about it. but we need something that looks into both input and context while evaluating the output. Like faithfulness for the same dataset above is 44.18604651162791%. I updated answer_relevance to do this since In my mind, faithfulness is faithfulness to the context. Most of the human eval for a RAG system to eval end to end would look at the input and the answer while considering the context. So we need one metric to replicate that? If we think it should be a new one I am happy to do that instead. Also in the correct answer_relevance, if we are not considering the context, is there a reason we have added `grading_context_columns = ["context"]` in the definition as required, that is more confusing to the users. If the context is required, the assumption is that it would be used. Let me know if I am missing something.

**Comment by annzhang-db** [comment]
> > Sorry, put this as a comment earlier - but I'm confused why we want to do this. We originally introduced relevance (to context) [here](https://github.com/mlflow/mlflow/pull/9724), but intentionally split out metrics into answer relevance (which only considers the input) and faithfulness (which only considers the context) to be more consistent with RAGAS: https://docs.ragas.io/en/latest/concepts/metrics/answer_relevance.html. Were the instructions for human annotators to consider relevance to context (the definition may be different than what we are looking for with this metric)? In the case that we are adding this, we are blurring the lines between faithfulness and answer relevance and it gets more ambiguous.
> 
> Yup I thought about it. but we need something that looks into both input and context while evaluating the output. Like faithfulness for the same dataset above is 44.18604651162791%. I updated answer_relevance to do this since In my mind, faithfulness is faithfulness to the context. Most of the human eval for a RAG system to eval end to end would look at the input and the answer while considering the context. So we need one metric to replicate that? If we think it should be a new one I am happy to do that instead. Also in the correct answer_relevance, if we are not considering the context, is there a reason we have added `grading_context_columns = ["context"]` in the definition as required, that is more confusing to the users. If the context is required, the assumption is that it would be used. Let me know if I am missing something.

Sorry, there shouldn't be "context" here - this is my mistake (I'll put out a PR to fix it). I think if we want to add a metric for relevance to BOTH input and output we should introduce a new metric OR we should rename it to be different from RAGAS so it is clear to users that we are using a different definition than RAGAS.

**Pull request closed without merging**